cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -473,10 +473,16 @@ public RunRecordMeta recordProgramDeprovisioning(ProgramRunId programRunId, byte
     ProgramRunClusterStatus clusterStatus = existing.getCluster().getStatus();
     delete(existing);
     key.add(getInvertedTsKeyPart(existing.getStartTs())).add(programRunId.getRun()).build();
+    ProgramRunStatus status = existing.getStatus();
 
-    // if the previous state was provisioning, that means we've transitioned here from a failure
-    ProgramRunStatus newStatus = clusterStatus == ProgramRunClusterStatus.PROVISIONING ?
-      ProgramRunStatus.FAILED : existing.getStatus();
+    // if the previous state was provisioning, that means we've transitioned here from a failure or from a stop
+    // if we've transitioned here from a stop, the program run status will be killed.
+    // if we've transitioned here from a failure, the program run status will be pending,
+    // which needs to be transitioned to failed.
+    ProgramRunStatus newStatus = existing.getStatus();
+    if (clusterStatus == ProgramRunClusterStatus.PROVISIONING && status == ProgramRunStatus.PENDING) {
```

Comment:
Would be good to rename `status` as `programStatus`, and `newStatus` as `newProgramStatus` to differentiate it from `clusterStatus`.

Reply:
done